lls/main.c

```diff
@@ -16,11 +16,206 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include <rte_ethdev.h>
+#include <rte_malloc.h>
+
 #include "gatekeeper_lls.h"
+#include "gatekeeper_main.h"
+#include "gatekeeper_net.h"
+#include "arp.h"
+
+/*
+ *	ARP
+ */
+
+static int
+modify_arp_filter(uint8_t port_id, uint16_t queue_id, enum rte_filter_op op)
+{
+	int ret;
+	struct rte_eth_ethertype_filter filter;
+
+	/* L2 filter for ARP packets. */
+	memset(&filter, 0, sizeof(filter));
+	filter.ether_type = rte_cpu_to_be_16(ETHER_TYPE_ARP);
+	filter.queue = queue_id;
+
+	ret = rte_eth_dev_filter_ctrl(port_id, RTE_ETH_FILTER_ETHERTYPE,
+		op, &filter);
+	if (ret < 0)
+		RTE_LOG(ERR, PORT,
+			"Could not modify (op=%hhu) ARP filter on port %hhu!",
+			op, port_id);
+	return ret;
+}
+
+static int
+start_arp(void)
+{
+	struct net_config *net_conf = get_net_conf();
+	uint16_t queue_id = get_queue_id();
+	int ret;
+
+	ret = modify_arp_filter(net_conf->front.id, queue_id,
+		RTE_ETH_FILTER_ADD);
+	if (ret < 0)
+		return ret;
+
+	ret = modify_arp_filter(net_conf->back.id, queue_id,
+		RTE_ETH_FILTER_ADD);
+	if (ret < 0)
+		goto front;
+
+	ret = arp_cache_init();
+	if (ret < 0)
+		goto back;
+
+	return 0;
+back:
+	modify_arp_filter(net_conf->back.id, queue_id, RTE_ETH_FILTER_DELETE);
+front:
+	modify_arp_filter(net_conf->front.id, queue_id, RTE_ETH_FILTER_DELETE);
+	return ret;
+}
+
+static void
+stop_arp(void)
+{
+	struct net_config *net_conf = get_net_conf();
+	arp_cache_destroy();
+	modify_arp_filter(net_conf->back.id, (uint16_t)rte_lcore_id(),
+		RTE_ETH_FILTER_DELETE);
+	modify_arp_filter(net_conf->front.id, (uint16_t)rte_lcore_id(),
```

Comment:
`get_queue_id()` should be used here, not `rte_lcore_id()` in the two cases above.